jcstress-samples/src/main/java/org/openjdk/jcstress/samples/concurreny/mutex/Mutex_02_PetersonAlgorithm.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.jcstress.samples.concurreny.mutex;
+
+import org.openjdk.jcstress.annotations.Actor;
+import org.openjdk.jcstress.annotations.JCStressTest;
+import org.openjdk.jcstress.annotations.Outcome;
+import org.openjdk.jcstress.annotations.State;
+import org.openjdk.jcstress.infra.results.II_Result;
+import org.openjdk.jcstress.infra.results.ZZ_Result;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE;
+import static org.openjdk.jcstress.annotations.Expect.FORBIDDEN;
+
+/*
+    How to run this test:
+        $ java -jar jcstress-samples/target/jcstress.jar -t Mutex_02_PetersonAlgorithm
+ */
+
+/**
+ * Implemented according to https://en.wikipedia.org/wiki/Peterson%27s_algorithm
+ */
+@JCStressTest
+@Outcome(id = {"false, false"}, expect = ACCEPTABLE, desc = "Both actors have entered the critical section one after another")
+@Outcome(expect = FORBIDDEN, desc = "Both actors have entered the critical section at the same time")
+@State
+public class Mutex_02_PetersonAlgorithm {
+    private final AtomicBoolean flagForActor1 = new AtomicBoolean();
+    private final AtomicBoolean flagForActor2 = new AtomicBoolean();
+    private final AtomicInteger turn = new AtomicInteger();
+    private volatile boolean taken1, taken2;
```

Comment:
`flagForActor1`, `flagForActor2`, and `turn` should be volatiles, taken1 and taken2 should be regular fields.

Reply:
Can you explain me why taken1 and taken2 must not be volatile?
I mean actor1 and actor2 run in different threads, don't they?
If so, actor1 won't see the current value of taken2 when actor2 changes taken2 or am I wrong?

Reply:
Writes to plain, non-volatile fields can be observed by other threads as well.

In case you make `taken1` and `taken2` volatiles, you might accidentally re-create IRIW case out of these `taken1`, `taken2`, and a couple of other fields (see https://github.com/openjdk/jcstress/blob/df83b446f187ae0b0fa31fa54decb59db9e955da/tests-custom/src/main/java/org/openjdk/jcstress/tests/volatiles/VolatileIRIWTest.java )

In other words, `17.4.4. Synchronization Order` reads `A synchronization order is a total order over all of the synchronization actions of an execution`, while `A write to a volatile variable v (§8.3.1.4) synchronizes-with all subsequent reads of v by any thread`. That means, if you make `takenX` volatile, you might create new `synchronizes-with` edges thus forbidding program behaviors.

Of course, it might be the case that writes to `taken1` and `taken2` is **never** visible to the other threads (I doubt compilers would be able to do such proof soon), so to play safe, both cases needs to be verified: non-volatile `taken1,2` and making one of them volatile.

Reply:
Ah thanks for that answer. I have not got such deep insights into the "secondary effects" of volatile.
I have removed it from all 4 samples.
Just another question: Is it theoretically possible that we get false positives without volatile?
I mean taken1 and taken2 are updated by their actors but the other actor might not see it.
Then both actors might enter the critical section at the same time without any actor noticing it in the results.
Of course, I know there is no free lunch.

Reply:
Just do this: https://github.com/openjdk/jcstress/pull/85#discussion_r660633790